### DISCUSSION TITLE
How to query and access resources in exclusive world query

### DISCUSSION THREAD

**Discussion by feelingsonice** [bevyengine/bevy#11587] [Q&A]
Please correct me if I'm wrong, but it seems like the only way to query for anything when using exclusive world queries is through the use of a mutable borrow, something like this:
```rust
fn exclusive(world: &mut World) {
    let mut query = world.query::<Entity>();
    for entity in query.iter(world) { // requires mutable access to world?
         //...
    }
}
```

The method signature of `QueryState::iter()` is indeed `pub fn iter<'w, 's>(&'s mut self, world: &'w World)`.

So..., this means that you cannot access any other readonly data while this is happening. For me, I need to access a `Resource` while iterating the query, but I can't...
```rust
fn exclusive(world: &mut World) {
    let resource = world.resource::<MyResource>(); // borrow checker refuses to allow this
    let mut query = world.query::<Entity>();
    for entity in query.iter(world) { // mutable borrow occurs
         //...
    }
}
```

So how does one go about doing something like this considering it's a common use case?

**Comment by nicopap** [comment]
You would use the [`SystemState`](https://docs.rs/bevy/latest/bevy/ecs/system/struct.SystemState.html#example) API.

It allows accessing system parameters similarly to a system, but using a `&mut World`.

```rust
let mut system_state: SystemState<(
    Query<(Entity, &mut Transform)>,
    Query<&mut Visibility>,
)> = SystemState::new(world);

let (mut transforms, mut visibs) = system_state.get_mut(world);

for (entity, mut tran) in &mut transforms {
    let Ok(mut vis) = visibs.get_mut(entity) else {
        continue;
    };
    // ... 
}
```
